internal/storage/gcs/bucket.go

```diff
@@ -20,6 +20,39 @@ import (
 	"golang.org/x/net/context"
 )
 
+// ChunkUploader interface represents asynchronous object
+// uploaders which upload a chunk of an object's data in a
+// call using GCS' resumable upload API.
+//
+// On closing, they return a GCS object (gcs.Object)
+// and and error object for error-handling.
+type ChunkUploader interface {
+	// UploadChunkAsync writes the entire input content by reading
+	// from the given reader interface.
+	//
+	// io.EOF is an expected error in case the reader has less
+	// data than the writer tries to read from it. When EOF is
+	// returned and whether it's an actionable error, depends on
+	// the actual implementation and invocation of this interface.
+	// In other words, EOF might not be a fatal error in some cases.
+	//
+	// Unlike the io.Writer interface, it doesn't return number-of-bytes
+	// uploaded in this call, as the write is asynchronous; instead
+	// this interface instead has
+	// BytesWrittenSoFar function below, which returns the total number
+	// of bytes successfully uploaded so far.
+	UploadChunkAsync(contents io.Reader) error
+
+	// Close closes out the writing of the GCS object and returns
```

Comment:
Close finalizes the upload and returns the created object. Error is returned in case of failures.

Reply:
Sure, changed to your write-up.